presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java

```diff
@@ -0,0 +1,224 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class FunctionManager
+{
+    private static final String TEMP_DEFAULT_CATALOG = "system";
+    private static final String TEMP_DEFAULT_SCHEMA = "functions";
+
+    private volatile Map<String, FunctionNamespace> functionNamespaces;
+    private final FunctionNamespace operatorNamespace;
+    private final TypeManager typeManager;
+    private final BlockEncodingSerde blockEncodingSerde;
+    private final FeaturesConfig featuresConfig;
+
+    public FunctionManager(TypeManager typeManager, BlockEncodingSerde blockEncodingSerde, FeaturesConfig featuresConfig)
+    {
+        functionNamespaces = ImmutableMap.of();
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.featuresConfig = requireNonNull(featuresConfig, "featuresConfig is null");
+        operatorNamespace = new StaticFunctionNamespace(typeManager, blockEncodingSerde, featuresConfig);
+    }
+
+    public void addFunctions(List<? extends SqlFunction> functions)
+    {
+        addFunctions(TEMP_DEFAULT_CATALOG, functions);
+    }
+
+    public synchronized void addFunctions(String catalog, List<? extends SqlFunction> functions)
+    {
+        Map<Boolean, List<SqlFunction>> operatorAndFunctionMap = functions.stream()
+                .collect(Collectors.partitioningBy(function -> isOperator(function.getSignature())));
+
+        operatorNamespace.addFunctions(operatorAndFunctionMap.get(true));
+
+        if (!functionNamespaces.containsKey(catalog)) {
+            FunctionNamespace namespace = new StaticFunctionNamespace(typeManager, blockEncodingSerde, featuresConfig);
+            namespace.addFunctions(operatorAndFunctionMap.get(false));
+            functionNamespaces = ImmutableMap.<String, FunctionNamespace>builder()
+                    .putAll(functionNamespaces)
+                    .put(catalog, namespace)
+                    .build();
+        }
+        else {
+            functionNamespaces.get(catalog).addFunctions(operatorAndFunctionMap.get(false));
+        }
+    }
+
+    public List<SqlFunction> list(Session session)
+    {
+        ImmutableList.Builder<SqlFunction> builder = ImmutableList.builder();
+        builder.addAll(operatorNamespace.list());
+        for (SqlPathElement element : session.getPath().getParsedPath()) {
+            FunctionNamespace namespace = functionNamespaces.get(getCatalog(element, session));
+            if (namespace != null) {
+                builder.addAll(namespace.list());
+            }
+        }
+        return builder.build();
+    }
+
+    public Signature resolveFunction(Session session, QualifiedName name, List<TypeSignatureProvider> parameterTypes)
+    {
+        try {
+            return operatorNamespace.resolveFunction(name, parameterTypes);
```

Comment:
I think we should have two methods: `resolveFunction` and `resolveOperator` instead of trying to do both in the same method.